File: tests/src/test/native/com/example/native_value_profile_fuzzer.cpp
// Copyright 2024 Code Intelligence GmbH
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <cstdint>
#include <cstring>

#include "com_example_NativeValueProfileFuzzer.h"

// Prevent the compiler from inlining the secret all the way into checkAccess,
// which would make it trivial for the fuzzer to pass the checks.
volatile uint64_t secret = 0xefe4eb93215cb6b0L;

static uint64_t insecureEncrypt(uint64_t input) { return input ^ secret; }

jboolean Java_com_example_NativeValueProfileFuzzer_checkAccess(JNIEnv *, jclass,
                                                               jlong block1,
                                                               jlong block2) {
  if (insecureEncrypt(block1) == 0x9fc48ee64d3dc090L) {
    if (insecureEncrypt(block2) == 0x888a82ff483ad9c2L) {
      return true;
    }
  }
  return false;
}

File: tests/src/test/java/com/example/NativeValueProfileFuzzer.java
/*
 * Copyright 2024 Code Intelligence GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example;

import com.code_intelligence.jazzer.api.FuzzedDataProvider;
import com.code_intelligence.jazzer.api.FuzzerSecurityIssueLow;
import com.github.fmeum.rules_jni.RulesJni;

public class NativeValueProfileFuzzer {
  public static void fuzzerInitialize() {
    RulesJni.loadLibrary("native_value_profile_fuzzer", NativeValueProfileFuzzer.class);
  }

  public static void fuzzerTestOneInput(FuzzedDataProvider data) {
    long[] blocks = data.consumeLongs(2);
    if (blocks.length != 2) return;
    if (checkAccess(blocks[0], blocks[1])) {
      throw new FuzzerSecurityIssueLow("Security breached");
    }
  }

  private static native boolean checkAccess(long block1, long block2);
}
